ql/src/java/org/apache/hadoop/hive/ql/optimizer/GroupingSetOptimizer.java

```diff
@@ -0,0 +1,362 @@
+          candidates.add(((ExprNodeColumnDesc) key).getColumn());
+        }
+      }
+      candidates.retainAll(colNamesInSignature);
+
+      List<ColStatistics> columnStatistics = new ArrayList<>(parentOp.getStatistics().getColumnStats());;
+      columnStatistics.sort(new Comparator<ColStatistics>() {
+        @Override
+        public int compare(ColStatistics o1, ColStatistics o2) {
+          if (o1.getCountDistint() == o2.getCountDistint()) {
+            return 0;
+          } else {
+            // sort in reversed order, i.e., the largest comes the first
+            return (o1.getCountDistint() < o2.getCountDistint()) ? 1 : -1;
+          }
+        }
+      });
+
+      String partitionCol = null;
+      for (ColStatistics col: columnStatistics) {
+        String colName = col.getColumnName();
+        if (parentOp.getColumnExprMap().containsKey(colName) && candidates.contains(colName)) {
+          partitionCol = colName;
+          break;
+        }
+      }
+
+      if (partitionCol == null) {
+        LOG.debug("Skip grouping-set optimization as there is no feasible column in parent operator {}.",
+            parentOp);
+      }
+
+      return partitionCol;
+    }
+
+    private ReduceSinkOperator createReduceSink(Operator<?> parentOp, String partitionColName,
+        GroupingSetProcessorContext context) {
+      Map<String, ExprNodeDesc> colExprMap = new HashMap<>();
+      List<ExprNodeDesc> valueColumns = new ArrayList<>();
+      List<String> valueColumnNames = new ArrayList<>();
+      List<ColumnInfo> signature = new ArrayList<>();
+      List<ExprNodeDesc> partCols = new ArrayList<>();
+
+      for (ColumnInfo pColInfo: parentOp.getSchema().getSignature()) {
+        ColumnInfo cColInfo = new ColumnInfo(pColInfo);
+        String pColName = pColInfo.getInternalName();
+        valueColumnNames.add(pColName);
+
+        String cColName = Utilities.ReduceField.VALUE + "." + pColName;
+        cColInfo.setInternalName(cColName);
+        signature.add(cColInfo);
+
+        ExprNodeDesc valueExpr = new ExprNodeColumnDesc(pColInfo);
+        valueColumns.add(valueExpr);
+        colExprMap.put(cColName, valueExpr);
+
+        if (pColName.equals(partitionColName)) {
+          partCols.add(valueExpr);
+        }
+      }
+
+      List<FieldSchema> valueFields =
+          PlanUtils.getFieldSchemasFromColumnList(valueColumns, valueColumnNames, 0, "");
+      TableDesc valueTable = PlanUtils.getReduceValueTableDesc(valueFields);
+
+      TableDesc keyTable = PlanUtils.getReduceKeyTableDesc(new ArrayList<>(), "", "");
+
+      List<ExprNodeDesc> keyColumns = new ArrayList<>();
+      List<String> keyColumnNames = new ArrayList<>();
+      List<List<Integer>> distinctColumnIndices = new ArrayList<>();
+
+      ReduceSinkDesc rsConf = new ReduceSinkDesc(keyColumns, 0, valueColumns, keyColumnNames,
+          distinctColumnIndices, valueColumnNames, -1, partCols, -1, keyTable, valueTable,
```

Comment:
I remember the argument between `partCols` and `keyTable` is # of reducers. We can move L294.